tests/vnc/vnc_test.go

```diff
@@ -0,0 +1,218 @@
+package lim
+
+import (
+	"flag"
+	"fmt"
+	"github.com/docker/docker/pkg/namesgenerator"
+	"github.com/lf-edge/eden/pkg/defaults"
+	"github.com/lf-edge/eden/pkg/device"
+	"github.com/lf-edge/eden/pkg/expect"
+	"github.com/lf-edge/eden/pkg/projects"
+	"github.com/lf-edge/eden/pkg/utils"
+	"github.com/lf-edge/eve/api/go/info"
+	log "github.com/sirupsen/logrus"
+	"math/rand"
+	"net"
+	"os"
+	"testing"
+	"time"
+)
+
+// This test deploys the VM with image http://cdimage.debian.org/cdimage/openstack/current/debian-10.4.3-20200610-openstack-ARCH.qcow2
+// with ARCH from config and vncDisplay into EVE
+// waits for the RUNNING state and checks access to VNC console
+// and removes app from EVE
+
+var (
+	timewait     = flag.Int("timewait", 300, "Timewait for items waiting in seconds")
+	name         = flag.String("name", "", "Name of app, random if empty")
+	tc           *projects.TestContext
+	externalIP   string
+	externalPort int
+	appName      string
+	vncDisplay   = 0
+	metadata     = "#cloud-config\npassword: passw0rd\nchpasswd: { expire: False }\nssh_pwauth: True\n"
+	appLink      = func(arch string) string {
+		return fmt.Sprintf("http://cdimage.debian.org/cdimage/openstack/current/debian-10.4.3-20200610-openstack-%s.qcow2", arch)
+	}
+)
+
+// TestMain is used to provide setup and teardown for the rest of the
+// tests. As part of setup we make sure that context has a slice of
+// EVE instances that we can operate on. For any action, if the instance
+// is not specified explicitly it is assumed to be the first one in the slice
+func TestMain(m *testing.M) {
+	fmt.Println("VNC access to app Test")
+
+	tc = projects.NewTestContext()
+
+	projectName := fmt.Sprintf("%s_%s", "TestVNCAccess", time.Now())
+
+	tc.InitProject(projectName)
+
+	tc.AddEdgeNodesFromDescription()
+
+	tc.StartTrackingState(false)
+
+	res := m.Run()
+
+	os.Exit(res)
+}
+
+//getVNCPort calculate port for vnc
+//for qemu it is forwarded
+//for rpi it is direct
+func getVNCPort(edgeNode *device.Ctx, vncDisplay int) int {
+	if edgeNode.GetDevModel() == defaults.DefaultRPIModel {
+		return 5901 + vncDisplay
+	} else {
+		return 5911 + vncDisplay //forwarded by qemu ports
+	}
+}
+
+//checkAppRunning wait for info of ZInfoApp type with mention of deployed AppName and ZSwState_RUNNING state
+func checkAppRunning(appName string) projects.ProcInfoFunc {
+	return func(msg *info.ZInfoMsg) error {
+		if msg.Ztype == info.ZInfoTypes_ZiApp {
+			if msg.GetAinfo().AppName == appName {
+				if msg.GetAinfo().State == info.ZSwState_RUNNING {
+					return fmt.Errorf("app RUNNING with name %s", appName)
+				}
+			}
+		}
+		return nil
+	}
+}
+
+//getEVEIP wait for IPs of EVE and returns them
+func getEVEIP(edgeNode *device.Ctx) projects.ProcTimerFunc {
+	return func() error {
+		if edgeNode.GetRemote() {
+			if eveIPCIDR, err := tc.GetState(edgeNode).LookUp("Dinfo.Network[0].IPAddrs[0]"); err != nil {
```

Comment:
@sadov @mydatascience here you can see attempt to use lookup for field state access.